Zend/zend_enum.c

```diff
@@ -371,3 +371,138 @@ void zend_enum_register_props(zend_class_entry *ce)
 		zend_declare_typed_property(ce, ZSTR_KNOWN(ZEND_STR_VALUE), &value_default_value, ZEND_ACC_PUBLIC, NULL, value_type);
 	}
 }
+
+static const zend_function_entry unit_enum_methods[] = {
+	ZEND_NAMED_ME(cases, zend_enum_cases_func, arginfo_class_UnitEnum_cases, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
+	ZEND_FE_END
+};
+
+static const zend_function_entry backed_enum_methods[] = {
+	ZEND_NAMED_ME(cases, zend_enum_cases_func, arginfo_class_UnitEnum_cases, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
+	ZEND_NAMED_ME(from, zend_enum_from_func, arginfo_class_BackedEnum_from, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
+	ZEND_NAMED_ME(tryFrom, zend_enum_try_from_func, arginfo_class_BackedEnum_tryFrom, ZEND_ACC_PUBLIC | ZEND_ACC_STATIC)
+	ZEND_FE_END
+};
+
+ZEND_API zend_class_entry *zend_register_internal_enum(
+	const char *name, zend_uchar type, zend_function_entry *functions)
+{
+	ZEND_ASSERT(type == IS_UNDEF || type == IS_LONG || type == IS_STRING);
+
+	zend_class_entry tmp_ce;
+	INIT_CLASS_ENTRY_EX(tmp_ce, name, strlen(name), functions);
+
+	zend_class_entry *ce = zend_register_internal_class(&tmp_ce);
+	ce->ce_flags |= ZEND_ACC_ENUM;
+	ce->enum_backing_type = type;
+	if (type != IS_UNDEF) {
+		ce->backed_enum_table = pemalloc(sizeof(HashTable), 1);
+		zend_hash_init(ce->backed_enum_table, 0, NULL, ZVAL_PTR_DTOR, 1);
```

Comment:
If these will be mostly generated by stubs it would make sense to pass the right number of cases here right away.

Reply:
I think _mostly_ is the point here. I don't see any reason to make the API more cumbersome than necessary.